server/http_tunnel.go

```diff
@@ -27,9 +28,19 @@ func StartHTTPSTunnel(addr, externalProxy, certFile, keyFile, username, password
 	proxy := goproxy.NewProxyHttpServer()
 
 	if externalProxy != "" {
-		proxy.Tr = &http.Transport{Proxy: func(req *http.Request) (*url.URL, error) {
-			return url.Parse(externalProxy)
-		}}
+		u, err := url.Parse(externalProxy)
+		if err != nil {
+			log.Printf("HTTP(S) Tunnel: failed to parse external proxy: %s\n", err.Error())
+			return
```

Comment:
This `return` statement reminds me that this tunnel server might not be launched at all once there is an error here, and also there isn't any re-launch mechanism when I looked into the main function. Is this an expected behaviour?

Reply:
Good point.  My reasoning is that since this is a configuration error, we should not retry.  Maybe it's less confusing if we quit the entire program on this error.

Reply:
Turns out we already do that: https://github.com/PelionIoT/edge-proxy/blob/6deaaacf839ba7ce5b332da02ec68a561e1ac43f/cmd/edge-proxy/main.go#L88

Reply:
In that case, can you revert the syntax change for `Proxy`?

Reply:
@moonchen even though the configuration is okay, it is still possible that this HTTP tunnel server could exit itself at anytime, unless we allow this situation to happen (HTTP tunnel exits itself while edge-proxy is still running). Either way, we should add some recover mechanism when the HTTP tunnel server exits itself.

Reply:
If the URL doesn't parse, I'd expect this to keep failing even if we try to restart the tunnel.

Reply:
> In that case, can you revert the syntax change for `Proxy`?

Which change are you referring to?

Reply:
At a high level, my thought here is that this check is only for the sake of completeness.  This should never error in edge-proxy since we already checked it in main.go, and would have exited if it failed to parse.  In case someone refactors the code or reuses this function, it is possible they call this function with an incorrect `externalHTTPProxyURI`.  I'd like to handle the error.   Since this is a configuration error, we have no way to gracefully recover.  In that case, I feel that the sensible thing to do is to print an error for the user, and stop.

Reply:
I was suggesting to change the `Proxy` definition back to the way it was. I.e.,
```
Proxy: func(req *http.Request) (*url.URL, error) {
			return url.Parse(externalProxy)
	}
```
But I see now that you're reusing `u` in TLSClientConfig to pull out the Hostname.

In light of that, should the error check on line 32 be an assert() instead since the value was already checked in main.go?  Are there assertions in go?

Reply:
Go doesn't have assertions: https://golang.org/doc/faq#assertions

Reply:
> Go doesn't have assertions: [golang.org/doc/faq#assertions](https://golang.org/doc/faq#assertions)

@moonchen  @costanic Even though `go` doesn't have assertions, it has its own error handling mechanism to have some alternatives still.

Reply:
> At a high level, my thought here is that this check is only for the sake of completeness. This should never error in edge-proxy since we already checked it in main.go, and would have exited if it failed to parse.
> 
> In case someone refactors the code or reuses this function, it is possible they call this function with an incorrect `externalHTTPProxyURI`. I'd like to handle the error. Since this is a configuration error, we have no way to gracefully recover. In that case, I feel that the sensible thing to do is to print an error for the user, and stop.


@moonchen I got your point. But I don't think the current implementation could handle any error raised by `StartHTTPTunnel()` or `StartHTTPSTunnel()` as the way you think right now (no gracefully shutdown). Because [this](https://github.com/PelionIoT/edge-proxy/blob/1f34563e316ddf82632110049da8608a821bcdba/cmd/edge-proxy/main.go#L144) will block `edge-proxy` exits itself.

Reply:
Maybe this is a defect rather than the relative changes that required from this PR. It might be great to have a separate PR to have the appropriate recover mechanism.